src/test/pages/Gameboard.test.tsx

```diff
@@ -0,0 +1,144 @@
+        u.role = role;
+        u.id = 301;
+      }),
+    PageComponent: Gameboard,
+    initialRouteEntries: ["/gameboard#test-gameboard-1"],
+    extraEndpoints: [
+      rest.get(API_PATH + "/gameboards/test-gameboard-1", (req, res, ctx) => {
+        return res(ctx.status(200), ctx.json(mockGameboard));
+      }),
+      rest.post(API_PATH + "/gameboards/test-gameboard-1", (req, res, ctx) => {
+        return res(ctx.status(200), ctx.json({ ...mockGameboard, title: req.url.searchParams.get("title") }));
+      }),
+    ],
+  });
+  return mockGameboard;
+};
+
+describe("Gameboard", () => {
+  it("renders expected title", async () => {
+    const gameboard = renderGameboard({ role: "STUDENT" });
+    await waitFor(() => {
+      expect(screen.queryByText("Loading...")).toBeNull();
+    });
+    const title = screen.getByRole("heading", { name: gameboard.title });
+    expect(title).toBeInTheDocument();
+  });
+
+  const teachersAndAbove: Role[] = USER_ROLES.filter((role) => role !== "STUDENT" && role !== "TUTOR");
+
+  it.each(teachersAndAbove)("offers gameboard title edit button for %s if they own the gameboard", async (role) => {
+    renderGameboard({ role });
+    await waitFor(() => {
+      expect(screen.queryByText("Loading...")).toBeNull();
+    });
+    const editButton = screen.getByRole("button", { name: "Edit" });
+    expect(editButton).toBeInTheDocument();
+  });
+
+  it.each(["STUDENT", "TUTOR"] as Role[])("does not offer gameboard title edit button for %s", async (role) => {
+    renderGameboard({ role });
+    await waitFor(() => {
+      expect(screen.queryByText("Loading...")).toBeNull();
+    });
+    const editButton = screen.queryByRole("button", { name: "Edit" });
+    expect(editButton).toBeNull();
+  });
+
+  it.each(USER_ROLES)(
+    "does not offer gameboard title edit button for %s if they do not own the gameboard",
+    async (role) => {
+      renderGameboard({ role, notGameboardOwner: true });
+      await waitFor(() => {
+        expect(screen.queryByText("Loading...")).toBeNull();
+      });
+      const editButton = screen.queryByRole("button", { name: "Edit" });
+      expect(editButton).toBeNull();
+    },
+  );
+
+  it("updates gameboard title if new title is submitted", async () => {
+    const mockSaveGameboard = jest.spyOn(isaacApi.endpoints.renameAndLinkUserToGameboard, "initiate");
+    const gameboard = renderGameboard({ role: "TEACHER" });
+    await waitFor(() => {
+      expect(screen.queryByText("Loading...")).toBeNull();
+    });
+    const editButton = screen.getByRole("button", { name: "Edit" });
+    await userEvent.click(editButton);
+    const mainHeading = screen.getByTestId("main-heading");
+    const inputField = within(mainHeading).getByRole("textbox");
+    await userEvent.type(inputField, "New Title");
+    const saveButton = within(mainHeading).getByRole("button", { name: "Save" });
+    await waitFor(async () => {
+      await userEvent.click(saveButton);
```

Comment:
This is interesting: how come this needs to be inside `waitFor`? Doesn't it work by calling it directly?

Reply:
I expected it to work just as `await userEvent.click(saveButton)` but it didn't like that. Pressing the button updates a state, and also triggers a post request to push the update to DB. When I made some other changes (like awaiting the new title) it then gave errors about state updates not being wrapped in act().  Pushed some changes now that seem to work without errors